<h1><u><span style="color:#3399FF;">Stroke Patient Healthcare Using Deep Learning</span></u></h1>


The primary aim of this project is to analyze a healthcare dataset focused on predicting **stroke** occurrences. This dataset includes patient details such as demographic information, medical history, and lifestyle factors. The objective is to estimate the likelihood of a stroke by examining various attributes.

<h4><u>Dataset:</u></h4>

- **Gender and Age**: Fundamental demographic details.
- **Hypertension and Heart Disease**: Medical history relevant to stroke risk.
- **Marital Status**: Indicates whether the individual has ever been married, potentially impacting health outcomes.
- **Employment Type and Residence**: Social factors that may influence health.
- **Average Glucose Level and BMI**: Health indicators linked to metabolic conditions.
- **Smoking Status**: Lifestyle factor associated with stroke risk.
- **Stroke**: The outcome variable, indicating if a stroke has occurred (1) or not (0).

**Dataset Link**: [Click here to access the dataset](https://drive.google.com/file/d/1XyhVIZaKYZczlM2alun_fofilqTBq_9c/view?usp=drive_link)


># **(1) Defining Problem Statement and Analyzing Basic Metrics**
***

This project aims to analyze a healthcare dataset specific to **stroke** patients, with a focus on identifying significant risk factors contributing to stroke occurrences. Key metrics include age, hypertension, heart disease, BMI, glucose levels, and lifestyle aspects such as smoking habits. By examining these factors, we seek to reveal patterns that support early diagnosis and stroke prevention. The analysis will offer insights into stroke prevalence across various demographics and health conditions, enhancing our understanding of contributing risk factors.

# **(2) Import libraries and Load the dataset**
***

In [14]:
import numpy as np  # Importing NumPy 
import pandas as pd  # Importing Pandas 


In [16]:
# Load .csv dataset into a pandas dataFrame
df=pd.read_csv("https://drive.google.com/uc?export=download&id=1XyhVIZaKYZczlM2alun_fofilqTBq_9c")

In [18]:
# shows the top 5 records of the dataset
df.head()

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


# **(3) Data Exploration and Pre-processing**
***

### **Review Basic Metrics and Data Types**

Understanding the dataset's structure, including the number of rows and columns and the data types of each attribute, is a crucial step. Exploring the data is essential for effective analysis.

In [24]:
df.shape  # Returns a tuple representing the number of rows and columns in the DataFrame

(5110, 12)

In [26]:
df.info()  # Provides a summary of the DataFrame, including data types and non-null counts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


<h5 style="color: blue;">Observations:</h5>

- **Dataset Dimensions**: Contains 5,110 entries and 12 columns.
- **Data Types**:
  - 4 columns with integer data
  - 3 columns with float data
  - 5 columns with object (categorical) data
- **Missing Values**: The **BMI** column has 101 missing entries; all other columns are fully populated.
- **Attributes**:
  - Demographic Details: **gender**, **age**, **ever_married**, **work_type**, **Residence_type**
  - Health Metrics: **hypertension**, **heart_disease**, **avg_glucose_level**, **bmi**, **smoking_status**, **stroke**
- **Analysis Scope**: The dataset is well-suited for examining the link between demographic factors and stroke risk.


In [31]:
 #Describing the statistical summary of numerical type data
df.describe()

,id,age,hypertension,heart_disease,avg_glucose_level,bmi,stroke
count,5110.000000,5110.000000,5110.000000,5110.000000,5110.000000,4909.000000,5110.000000
mean,36517.829354,43.226614,0.097456,0.054012,106.147677,28.893237,0.048728
std,21161.721625,22.612647,0.296607,0.226063,45.283560,7.854067,0.215320
min,67.000000,0.080000,0.000000,0.000000,55.120000,10.300000,0.000000
25%,17741.250000,25.000000,0.000000,0.000000,77.245000,23.500000,0.000000
50%,36932.000000,45.000000,0.000000,0.000000,91.885000,28.100000,0.000000
75%,54682.000000,61.000000,0.000000,0.000000,114.090000,33.100000,0.000000
max,72940.000000,82.000000,1.000000,1.000000,271.740000,97.600000,1.000000


<h5 style="color: blue;">Observations:</h5>

- **Age Range**: The participants have an average age of **43.23 years**, spanning from **0.08** to **82 years**, indicating a broad age distribution.
  
- **Prevalence of Hypertension and Heart Disease**: Approximately **9.75%** of participants have hypertension, and around **5.4%** have a history of heart disease, indicating relatively low prevalence rates.

- **Glucose Levels**: The mean glucose level is **106.15 mg/dL**, with a peak of **271.74 mg/dL**, showing significant variation among participants.

- **Body Mass Index (BMI)**: The average BMI is **28.89 kg/m²**, suggesting a tendency towards overweight in the sample; there are **201 missing values** in the BMI data.

- **Stroke Cases**: Around **4.87%** of participants have had a stroke, indicating a relatively low incidence rate within the dataset.


In [41]:
df.smoking_status.unique()  # Retrieve the unique values in the 'smoking_status' column to analyze different smoking behaviors in the dataset.

array(['formerly smoked', 'never smoked', 'smokes', 'Unknown'],
      dtype=object)

### **Statistical Summary of categorical type data**

In [44]:
# Statistical summary of categorical type data
df.describe(include = object)

,gender,ever_married,work_type,Residence_type,smoking_status
count,5110,5110,5110,5110,5110
unique,3,2,5,2,4
top,Female,Yes,Private,Urban,never smoked
freq,2994,3353,2925,2596,1892


<h5 style="color: blue;">Observations:</h5>

- **Gender**: The dataset contains **3 unique categories**, with **Female** being the most represented (2,994 occurrences), highlighting a greater presence of females.

- **Marital Status**: Most participants are **ever married** (3,353 occurrences), indicating that a significant portion of the dataset comprises individuals who have been married at some point.

- **Work Type**: There are **5 distinct work types**, with **Private** employment being the most prevalent (2,925 occurrences), suggesting that many participants are employed in the private sector.

- **Residence Type**: The dataset features **2 residence types**, with **Urban** being more common (2,596 occurrences), indicating a tendency for participants to reside in urban areas.

- **Smoking Status**: There are **4 unique categories** for smoking status, with **never smoked** being the most frequent (1,892 occurrences), suggesting a relatively low rate of smoking among participants.


<h3 style="text-decoration: underline;">Identifying Null Values</h3>

This phase involves both **data cleaning** and **data preprocessing**. Recognizing and handling missing values is a critical component of data cleaning, as it deals with the issue of incomplete data. Depending on the extent of the missing values, you may need to select a strategy for addressing them, such as imputing values or eliminating the affected rows or columns.

Moreover, this step is vital for data preprocessing because missing values can compromise the reliability of future analyses. Addressing these gaps ensures that the data is in the proper format for effective analysis.


In [48]:
# Display the count of missing values for each column
df.isnull().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

In [50]:
# Calculate the missing values percentage for each column and round to two decimal places
missing_values_percentage = (df.isnull().mean() * 100).round(2)

# Display the missing values percentage for each column
print("Missing Values Percentage:\n")
print(missing_values_percentage)

Missing Values Percentage:

id                   0.00
gender               0.00
age                  0.00
hypertension         0.00
heart_disease        0.00
ever_married         0.00
work_type            0.00
Residence_type       0.00
avg_glucose_level    0.00
bmi                  3.93
smoking_status       0.00
stroke               0.00
dtype: float64


<h5 style="color: blue;">Observations:</h5>

- The dataset is largely complete, with most columns showing **0% missing data**.
- The **'bmi'** column is the only one with missing values, accounting for **3.93%** of the entries.
- Key columns like **'id'**, **'gender'**, **'age'**, **'hypertension'**, **'heart_disease'**, **'ever_married'**,
  **'work_type'**, **'Residence_type'**, **'avg_glucose_level'**, **'smoking_status'**, and **'stroke'** have **no missing data**.


<h3 style="text-decoration: underline;">Handling null values</h3>



In [58]:
# Handling missing values for the 'bmi' column by assigning the result back
df['bmi'] = df['bmi'].fillna(df['bmi'].median())

In [60]:
# Display the count of missing values for each column
df.isnull().sum()

id                   0
gender               0
age                  0
hypertension         0
heart_disease        0
ever_married         0
work_type            0
Residence_type       0
avg_glucose_level    0
bmi                  0
smoking_status       0
stroke               0
dtype: int64

<h5 style="color: blue;">Observations:</h5>

- The dataset is complete with no missing values across all columns.